app/js/ZipManager.js

```diff
@@ -0,0 +1,69 @@
+const logger = require('logger-sharelatex')
+const UpdatesManager = require('./UpdatesManager')
+const DiffGenerator = require('./DiffGenerator')
+const DocumentUpdaterManager = require('./DocumentUpdaterManager')
+const PackManager = require('./PackManager')
+const yazl = require('yazl')
+const util = require('util')
+
+async function rewindDoc(projectId, docId, zipfile) {
+  logger.log({ projectId, docId }, 'rewinding document')
+  const [
+    finalContent,
+    version
+  ] = await DocumentUpdaterManager.promises.getDocument(projectId, docId)
+  zipfile.addBuffer(
+    Buffer.from(finalContent),
+    `${docId.toString()}/content/end/${version}`
+  )
+
+  // now rewind content
+  // TODO: retrieve updates incrementally
+  const updates = await PackManager.promises.getOpsByVersionRange(
+    projectId,
+    docId,
+    -1,
+    version
+  )
+
+  let content = finalContent
+  let v = version
+
+  for (const update of updates) {
+    zipfile.addBuffer(
+      Buffer.from(JSON.stringify(update)),
+      `${docId.toString()}/updates/${update.v}`,
+      { mtime: new Date(update.meta.start_ts) }
+    )
+    try {
+      content = DiffGenerator.rewindUpdate(content, update)
+      v = update.v
+    } catch (e) {
+      e.attempted_update = update // keep a record of the attempted update
+      logger.error({ projectId, docId, err: e }, 'rewind error')
+    }
+  }
+  zipfile.addBuffer(
+    Buffer.from(content),
+    `${docId.toString()}/content/start/${v}`
+  )
```

Comment:
Do you think it might be useful, at the end of writing the data for the doc, to also return a metadata object about the files that have been added to the archive?

That metadata could then be written to a `manifest.json` file at the root of the archive, which can then be used to fetch the files out of the archive without needing to parse information out of the file structure.

Reply:
Added in https://github.com/overleaf/track-changes/pull/118 (exact format of the manifest open to change, if needed)

Reply:
Good idea.